# ADS 509 Module 1: APIs and Web Scraping

This notebook has three parts. In the first part you will pull data from the Twitter API. In the second, you will scrape lyrics from AZLyrics.com. In the last part, you'll run code that verifies the completeness of your data pull. 

For this assignment you have chosen two musical artists who have at least 100,000 Twitter followers and 20 songs with lyrics on AZLyrics.com. In this part of the assignment we pull the some of the user information for the followers of your artist and store them in text files. 


## Important Note

This assignment requires you to have a version of Tweepy that is at least version 4. The latest version is 4.10 as I write this. Critically, this version of Tweepy is *not* on the upgrade path from Version 3, so you will not be able to simply upgrade the package if you are on Version 3. Instead you will need to explicitly install version 4, which you can do with a command like this: `pip install "tweepy>=4"`. You will also be using Version 2 of the Twitter API for this assignment. 

Run the below cell. If your version of Tweepy begins with a "4", then you should be good to go. If it begins with a "3" then run the following command, found [here](https://stackoverflow.com/questions/5226311/installing-specific-package-version-with-pip), at the command line or in a cell: `pip install -Iv tweepy==4.9`. (You may want to update that version number if Tweepy has moved on past 4.9. 

In [2]:
!pip3 install tweepy

## General Assignment Instructions

These instructions are included in every assignment, to remind you of the coding standards for the class. Feel free to delete this cell after reading it. 

One sign of mature code is conforming to a style guide. We recommend the [Google Python Style Guide](https://google.github.io/styleguide/pyguide.html). If you use a different style guide, please include a cell with a link. 

Your code should be relatively easy-to-read, sensibly commented, and clean. Writing code is a messy process, so please be sure to edit your final submission. Remove any cells that are not needed or parts of cells that contain unnecessary code. Remove inessential `import` statements and make sure that all such statements are moved into the designated cell. 

Make use of non-code cells for written commentary. These cells should be grammatical and clearly written. In some of these cells you will have questions to answer. The questions will be marked by a "Q:" and will have a corresponding "A:" spot for you. *Make sure to answer every question marked with a `Q:` for full credit.* 


# Twitter API Pull

In [2]:
# for the twitter section
import tweepy
import os
import datetime
import re
from pprint import pprint

# for the lyrics scrape section
import requests
import time
from bs4 import BeautifulSoup
from collections import defaultdict, Counter

In [2]:
# Use this cell for any import statements you add
from IPython.display import clear_output, display
import random
import time
import pandas as pd
import shutil
import os

We need bring in our API keys. Since API keys should be kept secret, we'll keep them in a file called `api_keys.py`. This file should be stored in the directory where you store this notebook. The example file is provided for you on Blackboard. The example has API keys that are _not_ functional, so you'll need to get Twitter credentials and replace the placeholder keys. 

In [3]:
from api_keys import api_key, api_key_secret, bearer_token

In [4]:
client = tweepy.Client(bearer_token,wait_on_rate_limit=True)

# Testing the API

The Twitter APIs are quite rich. Let's play around with some of the features before we dive into this section of the assignment. For our testing, it's convenient to have a small data set to play with. We will seed the code with the handle of John Chandler, one of the instructors in this course. His handle is `@37chandler`. Feel free to use a different handle if you would like to look at someone else's data. 

We will write code to explore a few aspects of the API: 

1. Pull some of the followers @37chandler.
1. Explore response data, which gives us information about Twitter users. 
1. Pull the last few tweets by @37chandler.


In [195]:
handle = "37chandler"

user_obj = client.get_user(username=handle)

followers = client.get_users_followers(user_obj.data.id, user_fields=["created_at","description","location", "public_metrics"])

Now let's explore these a bit. We'll start by printing out names, locations, following count, and followers count for these users. 

Let's find the person who follows this handle who has the most followers. 

In [191]:
max_followers = 0

for idx, user in enumerate(followers.data) :
    followers_count = user.public_metrics['followers_count']
    
    if followers_count > max_followers :
        max_followers = followers_count
        max_follower_user = user

        
print(max_follower_user)
print(max_follower_user.public_metrics)

WedgeLIVE
{'followers_count': 14157, 'following_count': 2223, 'tweet_count': 56071, 'listed_count': 218}


Let's pull some more user fields and take a look at them. The fields can be specified in the `user_fields` argument. 

In [192]:
response = client.get_user(id=user_obj.data.id,
                          user_fields=["created_at","description","location",
                                       "entities","name","pinned_tweet_id","profile_image_url",
                                       "verified","public_metrics"])

In [193]:
for field, value in response.data.items() :
    print(f"for {field} we have {value}")

for location we have MN
for verified we have False
for public_metrics we have {'followers_count': 193, 'following_count': 589, 'tweet_count': 994, 'listed_count': 3}
for id we have 33029025
for profile_image_url we have https://pbs.twimg.com/profile_images/2680483898/b30ae76f909352dbae5e371fb1c27454_normal.png
for created_at we have 2009-04-18 22:08:22+00:00
for username we have 37chandler
for description we have He/Him. Data scientist, urban cyclist, educator, erstwhile frisbee player. 

¯\_(ツ)_/¯
for name we have John Chandler


Now a few questions for you about the user object.

Q: How many fields are being returned in the `response` object? 

A: 4 fields (data, includes, errors, meta)

---

Q: Are any of the fields within the user object non-scalar? (I.e., more complicated than a simple data type like integer, float, string, boolean, etc.) 

A: Yes, the public_metrics is a json object which consists of multiple fields such as followers_count, following_count, etc.

---

Q: How many friends, followers, and tweets does this user have? 

A: 193 friends, 589 followers, and 994 tweets.


Although you won't need it for this assignment, individual tweets can be a rich source of text-based data. To illustrate the concepts, let's look at the last few tweets for this user. You are encouraged to explore the fields that are available about Tweets.

In [196]:

response = client.get_users_tweets(user_obj.data.id)

# By default, only the ID and text fields of each Tweet will be returned
for idx, tweet in enumerate(response.data) :
    print(tweet.id)
    print(tweet.text)
    print()
    
    if idx > 10 :
        break

1568242374085869570
RT @depthsofwiki: lots of memes about speedy wikipedia editors — quick thread about what went down on wikipedia in the minutes after her de…

1568074978754703361
@DrLaurenWilson @leighradwood @MaritsaGeorgiou @Walgreens I could not possibly agree more with this sentiment. Compared to almost any other primary care I've received, they are great.

1567530169686196224
@DrLaurenWilson @MaritsaGeorgiou @Walgreens For those who have access to Curry Health Center on campus, you can get a bivalent booster in 15 minutes from their delightful staff.

1567511181526708224
RT @shes_the_maNN1: I can’t describe how ancient this makes me feel. https://t.co/a1IvELjOFY

1567510612665864193
RT @AngryBlackLady: this is hilarious

1566031636457725953
RT @MarkJacob16: With all the arguments over whether MAGA Republicans are fascists, I reread William Shirer’s “The Rise and Fall of the Thi…

1563737816219000832
RT @wonderofscience: The Milky Way galaxy and a phenomenon known as "airglow" s

## Pulling Follower Information

In this next section of the assignment, we will pull information about the followers of your two artists. We've seen above how to pull a set of followers using `client.get_users_followers`. This function has a parameter, `max_results`, that we can use to change the number of followers that we pull. Unfortunately, we can only pull 1000 followers at a time, which means we will need to handle the _pagination_ of our results. 

The return object has the `.data` field, where the results will be found. It also has `.meta`, which we use to select the next "page" in the results using the `next_token` result. I will illustrate the ideas using our user from above. 


### Rate Limiting

Twitter limits the rates at which we can pull data, as detailed in [this guide](https://developer.twitter.com/en/docs/twitter-api/rate-limits). We can make 15 user requests per 15 minutes, meaning that we can pull $4 \cdot 15 \cdot 1000 = 60000$ users per hour. I illustrate the handling of rate limiting below, though whether or not you hit that part of the code depends on your value of `handle`.  


In the below example, I'll pull all the followers, 25 at a time. (We're using 25 to illustrate the idea; when you do this set the value to 1000.) 

In [217]:
handle_followers = []
pulls = 0
max_pulls = 5
next_token = None

while True :

    followers = client.get_users_followers(
        user_obj.data.id, 
        max_results = 25, 
        pagination_token = next_token,
        user_fields=["created_at","description","location",
                     "entities","name","pinned_tweet_id","profile_image_url",
                     "verified","public_metrics"]
    )
    pulls += 1
    
    for follower in followers.data : 
        follower_row = (follower.id,follower.name,follower.created_at,follower.description)
        handle_followers.append(follower_row)
    
    if 'next_token' in followers.meta and pulls < max_pulls :
        next_token = followers.meta['next_token']
    else : 
        break

## Pulling Twitter Data for Your Artists

Now let's take a look at your artists and see how long it is going to take to pull all their followers. 

In [68]:
artists = dict()

for handle in ['MaherZains','SamiYusuf'] : 
    user_obj = client.get_user(username=handle,user_fields=["public_metrics"])
    artists[handle] = (user_obj.data.id, 
                       handle,
                       user_obj.data.public_metrics['followers_count'])
    

for artist, data in artists.items() : 
    print(f"It would take {data[2]/(1000*15*4):.2f} hours to pull all {data[2]:,} followers for {artist}. \n")
    

It would take 4.84 hours to pull all 290,444 followers for MaherZains. 

It would take 12.79 hours to pull all 767,285 followers for SamiYusuf. 



Depending on what you see in the display above, you may want to limit how many followers you pull. It'd be great to get at least 200,000 per artist. 

As we pull data for each artist we will write their data to a folder called "twitter", so we will make that folder if needed.

In [37]:
# Make the "twitter" folder here. If you'd like to practice your programming, add functionality 
# that checks to see if the folder exists. If it does, then "unlink" it. Then create a new one.

if not os.path.isdir("twitter") : 
    #shutil.rmtree("twitter/")
    os.mkdir("twitter")

In this following cells, build on the above code to pull some of the followers and their data for your two artists. As you pull the data, write the follower ids to a file called `[artist name]_followers.txt` in the "twitter" folder. For instance, for Cher I would create a file named `cher_followers.txt`. As you pull the data, also store it in an object like a list or a data frame.

In addition to creating a file that only has follower IDs in it, you will create a file that includes user data. From the response object please extract and store the following fields: 

* screen_name	
* name	
* id	
* location	
* followers_count	
* friends_count	
* description

Store the fields with one user per row in a tab-delimited text file with the name `[artist name]_follower_data.txt`. For instance, for Cher I would create a file named `cher_follower_data.txt`. 

One note: the user's description can have tabs or returns in it, so make sure to clean those out of the description before writing them to the file. I've included some example code to do that below the stub. 

In [5]:
num_followers_to_pull = 200*1000 # feel free to use this to limit the number of followers you pull.

In [6]:
def convert_to_line(follower):
    
    user_id = str(follower.id)
    username = str(follower.username)
    name = str(follower.name)
    location = str(follower.location)
    followers_count = str(follower.public_metrics['followers_count'])
    following_count = str(follower.public_metrics['following_count'])
    
    description = follower.description
    description = description.replace('\\n', '|')
    description = description.replace("\n", "|") 
    description = description.replace("\t", "   ") 
    description = description.replace('"', '``')
    description = description.replace('\\u', ' ')     
    
    return '\t'.join([user_id, username, name, location, followers_count, following_count, description]) + "\n"
  

In [7]:
# Modify the below code stub to pull the follower IDs and write them to a file. 
fetch_fields = ['id', 'username', 'name', 'location', 'public_metrics', 'description']
column_names = ['user_id', 'username', 'name', 'location', 'followers_count', 'following_count', 'description']

handles = ['MaherZains','SamiYusuf']

# Grabs the time when we start making requests to the API
start_time = datetime.datetime.now()

for handle in handles :    
    followers_output_file = handle + "_followers.txt"
    user_data_output_file = handle + "_follower_data.txt"
    
    handle_followers = []
    next_token = '2VTJA8E6P27HAZZZ'
    count = 0
    user_obj = client.get_user(username=handle)
    
    with open(f'twitter/{user_data_output_file}','a') as f:
        f.write("\t".join(column_names) + '\n')
    
    with open(f'twitter/{followers_output_file}','a') as f:
        f.write('user_id' + '\n')    
    
    while True :

        followers = client.get_users_followers(
            user_obj.data.id,  max_results = 1000, 
            pagination_token = next_token, user_fields=fetch_fields)
                
        for follower in followers.data : 
            
            follower_info = convert_to_line(follower)
                        
            with open(f'twitter/{user_data_output_file}','a') as f:
                f.write(follower_info)
            
            with open(f'twitter/{followers_output_file}','a') as f:
                f.write(str(follower.id) + '\n')
                
            if count < num_followers_to_pull: 
                count += 1 
            else:
                break   
                                                                               
        if 'next_token' in followers.meta:
            next_token = followers.meta['next_token']
            
            with open(f'next_token_1.txt','a') as f:
                f.write(str(count) + ' ' + str(next_token) + '\n')     
                
        else:
            break
          
        
# Let's see how long it took to grab all follower IDs
end_time = datetime.datetime.now()
print(end_time - start_time)

Rate limit exceeded. Sleeping for 514 seconds.
Rate limit exceeded. Sleeping for 892 seconds.
Rate limit exceeded. Sleeping for 891 seconds.
Rate limit exceeded. Sleeping for 892 seconds.
Rate limit exceeded. Sleeping for 890 seconds.
Rate limit exceeded. Sleeping for 892 seconds.
Rate limit exceeded. Sleeping for 892 seconds.
Rate limit exceeded. Sleeping for 891 seconds.
Rate limit exceeded. Sleeping for 892 seconds.
Rate limit exceeded. Sleeping for 892 seconds.
Rate limit exceeded. Sleeping for 892 seconds.
Rate limit exceeded. Sleeping for 891 seconds.
Rate limit exceeded. Sleeping for 891 seconds.
Rate limit exceeded. Sleeping for 892 seconds.
Rate limit exceeded. Sleeping for 892 seconds.
Rate limit exceeded. Sleeping for 892 seconds.
Rate limit exceeded. Sleeping for 892 seconds.
Rate limit exceeded. Sleeping for 893 seconds.
Rate limit exceeded. Sleeping for 892 seconds.
Rate limit exceeded. Sleeping for 889 seconds.
Rate limit exceeded. Sleeping for 891 seconds.
Rate limit ex

12:39:44.472675


In [8]:
tricky_description = """
    Home by Warsan Shire
    
    no one leaves home unless
    home is the mouth of a shark.
    you only run for the border
    when you see the whole city
    running as well.

"""
# This won't work in a tab-delimited text file.

clean_description = re.sub(r"\s+"," ",tricky_description).strip()
clean_description

'Home by Warsan Shire no one leaves home unless home is the mouth of a shark. you only run for the border when you see the whole city running as well.'

---

# Lyrics Scrape

This section asks you to pull data from the Twitter API and scrape www.AZLyrics.com. In the notebooks where you do that work you are asked to store the data in specific ways. 

In [13]:
artists = {'maher_zain':"https://www.azlyrics.com/m/maherzain.html",
           'sami_yusuf':"https://www.azlyrics.com/s/samiyusuf.html"} 
# we'll use this dictionary to hold both the artist name and the link on AZlyrics

## A Note on Rate Limiting

The lyrics site, www.azlyrics.com, does not have an explicit maximum on number of requests in any one time, but in our testing it appears that too many requests in too short a time will cause the site to stop returning lyrics pages. (Entertainingly, the page that gets returned seems to only have the song title to [a Tom Jones song](https://www.azlyrics.com/lyrics/tomjones/itsnotunusual.html).) 

Whenever you call `requests.get` to retrieve a page, put a `time.sleep(5 + 10*random.random())` on the next line. This will help you not to get blocked. If you _do_ get blocked, which you can identify if the returned pages are not correct, just request a lyrics page through your browser. You'll be asked to perform a CAPTCHA and then your requests should start working again. 

## Part 1: Finding Links to Songs Lyrics

That general artist page has a list of all songs for that artist with links to the individual song pages. 

Q: Take a look at the `robots.txt` page on www.azlyrics.com. (You can read more about these pages [here](https://developers.google.com/search/docs/advanced/robots/intro).) Is the scraping we are about to do allowed or disallowed by this page? How do you know? 

A: Yes; since we are scraping a HTML file


In [29]:
def get_page_songs_hrefs(r):
    links = list()
    soup = BeautifulSoup(r.content, "html.parser") 
    listalbum = soup.findAll("div", class_ = "listalbum-item")

    def get_div_link(div):
        return BeautifulSoup(str(div), 'html.parser').find(href=True)['href']
    
    for div in listalbum:
        try:
            link = get_div_link(div)
            links.append(link)
        except:
            print(f'{div.string} has no link associated with it')
        
    return links   

# Let's set up a dictionary of lists to hold our links
lyrics_pages = defaultdict(list)


for artist, artist_page in artists.items() :
    r = requests.get(artist_page)
    lyrics_pages[artist] = get_page_songs_hrefs(r)
    time.sleep(5 + 10*random.random())
    
lyrics_pages

{'maher_zain': ['/lyrics/maherzain/alwaysbethere.html',
  '/lyrics/maherzain/yanabisalamalayka.html',
  '/lyrics/maherzain/inshaallah.html',
  '/lyrics/maherzain/palestinewillbefree.html',
  '/lyrics/maherzain/thankyouallah.html',
  '/lyrics/maherzain/allahiallahkiyakaro.html',
  '/lyrics/maherzain/thechosenone.html',
  '/lyrics/maherzain/barakaallahulakuma.html',
  '/lyrics/maherzain/fortherestofmylife.html',
  '/lyrics/maherzain/holdmyhand.html',
  '/lyrics/maherzain/awaken.html',
  '/lyrics/maherzain/subhanaallah.html',
  '/lyrics/maherzain/openyoureyes.html',
  '/lyrics/maherzain/yanabisalamalaykaarabicversion.html',
  '/lyrics/maherzain/iloveyouso.html',
  '/lyrics/maherzain/numberoneforme.html',
  '/lyrics/maherzain/mawlaya.html',
  '/lyrics/maherzain/mylittlegirl.html',
  '/lyrics/maherzain/forgiveme.html',
  '/lyrics/maherzain/onebigfamily.html',
  '/lyrics/maherzain/assalamualayka.html',
  '/lyrics/maherzain/paradise.html',
  '/lyrics/maherzain/mashaallah.html',
  '/lyrics/mah

Let's make sure we have enough lyrics pages to scrape. 

In [30]:
for artist, lp in lyrics_pages.items() :
    assert(len(set(lp)) > 20) 

In [31]:
# Let's see how long it's going to take to pull these lyrics 
# if we're waiting `5 + 10*random.random()` seconds 
for artist, links in lyrics_pages.items() : 
    print(f"For {artist} we have {len(links)}.")
    print(f"The full pull will take for this artist will take {round(len(links)*10/3600,2)} hours.")

For maher_zain we have 113.
The full pull will take for this artist will take 0.31 hours.
For sami_yusuf we have 114.
The full pull will take for this artist will take 0.32 hours.


## Part 2: Pulling Lyrics

Now that we have the links to our lyrics pages, let's go scrape them! Here are the steps for this part. 

1. Create an empty folder in our repo called "lyrics". 
1. Iterate over the artists in `lyrics_pages`. 
1. Create a subfolder in lyrics with the artist's name. For instance, if the artist was Cher you'd have `lyrics/cher/` in your repo.
1. Iterate over the pages. 
1. Request the page and extract the lyrics from the returned HTML file using BeautifulSoup.
1. Use the function below, `generate_filename_from_url`, to create a filename based on the lyrics page, then write the lyrics to a text file with that name. 


In [38]:
def generate_filename_from_link(link) :
    
    if not link :
        return None
    
    # drop the http or https and the html
    name = link.replace("https","").replace("http","")
    name = link.replace(".html","")

    name = name.replace("/lyrics/","")
    
    # Replace useless chareacters with UNDERSCORE
    name = name.replace("://","").replace(".","_").replace("/","_")
    
    # tack on .txt
    name = name + ".txt"
    
    return(name)

In [33]:
# Make the lyrics folder here. If you'd like to practice your programming, add functionality 
# that checks to see if the folder exists. If it does, then use shutil.rmtree to remove it and create a new one.

def create_folder(path):
    if not os.path.exists(path):
        os.makedirs(path)

create_folder('lyrics')

In [36]:
def get_song_lyrics_per_url(url):
    r = requests.get(url)
    soup = BeautifulSoup(r.content, "html.parser") 
    all_text = soup.find("div", class_ = "col-xs-12 col-lg-8 text-center") 
    starting_position = '<!-- Usage'
    end_position = '</div>'
    text_lines = str(all_text).splitlines()

    while True:
        if not starting_position in text_lines[0]:
            text_lines.pop(0)
        else:
            text_lines.pop(0)
            break
        
    text_lines = [line.replace('<br/>', '') for line in text_lines]
    
    lines_to_keep = list()
    for line in text_lines:
        if not end_position in line:
            lines_to_keep.append(line)
        else:
            break    
    
    return '\n'.join(lines_to_keep)

In [49]:
url_stub = "https://www.azlyrics.com" 
start = time.time()

for artist in lyrics_pages :
    create_folder(f'lyrics/{artist}')
    total_pages = 0 
    for lyrics_page in lyrics_pages[artist]:
        url = url_stub + lyrics_page
        song_lyrics = get_song_lyrics_per_url(url)
        file_name = generate_filename_from_link(lyrics_page)
        
        with open(f'lyrics/{artist}/{file_name}', 'w') as file:
            file.write(song_lyrics)
        
        print(f'{file_name} was saved under {artist}')
        total_pages += 1
        
        if total_pages == 25:
            break
        
        time.sleep(5 + 15 * random.random())
            

maherzain_alwaysbethere.txt was saved under maher_zain
maherzain_yanabisalamalayka.txt was saved under maher_zain
maherzain_inshaallah.txt was saved under maher_zain
maherzain_palestinewillbefree.txt was saved under maher_zain
maherzain_thankyouallah.txt was saved under maher_zain
maherzain_allahiallahkiyakaro.txt was saved under maher_zain
maherzain_thechosenone.txt was saved under maher_zain
maherzain_barakaallahulakuma.txt was saved under maher_zain
maherzain_fortherestofmylife.txt was saved under maher_zain
maherzain_holdmyhand.txt was saved under maher_zain
maherzain_awaken.txt was saved under maher_zain
maherzain_subhanaallah.txt was saved under maher_zain
maherzain_openyoureyes.txt was saved under maher_zain
maherzain_yanabisalamalaykaarabicversion.txt was saved under maher_zain
maherzain_iloveyouso.txt was saved under maher_zain
maherzain_numberoneforme.txt was saved under maher_zain
maherzain_mawlaya.txt was saved under maher_zain
maherzain_mylittlegirl.txt was saved under mah

In [50]:
print(f"Total run time was {round((time.time() - start)/3600,2)} hours.")

Total run time was 9.33 hours.


---

# Evaluation

This assignment asks you to pull data from the Twitter API and scrape www.AZLyrics.com.  After you have finished the above sections , run all the cells in this notebook. Print this to PDF and submit it, per the instructions.

In [6]:
# Simple word extractor from Peter Norvig: https://norvig.com/spell-correct.html
def words(text): 
    return re.findall(r'\w+', text.lower())

---

## Checking Twitter Data

The output from your Twitter API pull should be two files per artist, stored in files with formats like `cher_followers.txt` (a list of all follower IDs you pulled) and `cher_followers_data.txt`. These files should be in a folder named `twitter` within the repository directory. This code summarizes the information at a high level to help the instructor evaluate your work. 

In [43]:
twitter_files = os.listdir("twitter")
twitter_files = [f for f in twitter_files if f != ".DS_Store"]
artist_handles = list(set([name.split("_")[0] for name in twitter_files]))

print(f"We see two artist handles: {artist_handles[0]} and {artist_handles[1]}.")

We see two artist handles: SamiYusuf and MaherZains.


In [7]:
for artist in artist_handles :
    follower_file = artist + "_followers.txt"
    follower_data_file = artist + "_follower_data.txt"
    
    ids = open("twitter/" + follower_file,'r').readlines()
    
    print(f"We see {len(ids)-1} in your follower file for {artist}, assuming a header row.")
    
    with open("twitter/" + follower_data_file,'r') as infile :
        
        # check the headers
        headers = infile.readline().split("\t")
        
        print(f"In the follower data file ({follower_data_file}) for {artist}, we have these columns:")
        print(" : ".join(headers))
        
        description_words = []
        locations = set()
        
        
        for idx, line in enumerate(infile.readlines()) :
            line = line.strip("\n").split("\t")
            
            try : 
                locations.add(line[3])            
                description_words.extend(words(line[6]))
            except :
                pass
    
    
        print(f"We have {idx+1} data rows for {artist} in the follower data file.")

        print(f"For {artist} we have {len(locations)} unique locations.")

        print(f"For {artist} we have {len(description_words)} words in the descriptions.")
        print("Here are the five most common words:")
        print(Counter(description_words).most_common(5))

        print("")
        print("-"*40)
        print("")
    

We see 201931 in your follower file for SamiYusuf, assuming a header row.
In the follower data file (SamiYusuf_follower_data.txt) for SamiYusuf, we have these columns:
user_id : username : name : location : followers_count : following_count : description

We have 201931 data rows for SamiYusuf in the follower data file.
For SamiYusuf we have 20146 unique locations.
For SamiYusuf we have 832924 words in the descriptions.
Here are the five most common words:
[('none', 113941), ('i', 8758), ('and', 6116), ('of', 5879), ('a', 5591)]

----------------------------------------

We see 204254 in your follower file for MaherZains, assuming a header row.
In the follower data file (MaherZains_follower_data.txt) for MaherZains, we have these columns:
user_id : username : name : location : followers_count : following_count : description

We have 204264 data rows for MaherZains in the follower data file.
For MaherZains we have 27419 unique locations.
For MaherZains we have 742250 words in the descri

## Checking Lyrics 

The output from your lyrics scrape should be stored in files located in this path from the directory:
`/lyrics/[Artist Name]/[filename from URL]`. This code summarizes the information at a high level to help the instructor evaluate your work. 

In [8]:
artist_folders = os.listdir("lyrics/")
artist_folders = [f for f in artist_folders if os.path.isdir("lyrics/" + f)]

for artist in artist_folders : 
    artist_files = os.listdir("lyrics/" + artist)
    artist_files = [f for f in artist_files if 'txt' in f or 'csv' in f or 'tsv' in f]

    print(f"For {artist} we have {len(artist_files)} files.")

    artist_words = []

    for f_name in artist_files : 
        with open("lyrics/" + artist + "/" + f_name) as infile : 
            artist_words.extend(words(infile.read()))

            
    print(f"For {artist} we have roughly {len(artist_words)} words, {len(set(artist_words))} are unique.")


For sami_yusuf we have 25 files.
For sami_yusuf we have roughly 5314 words, 1330 are unique.
For maher_zain we have 25 files.
For maher_zain we have roughly 8659 words, 1235 are unique.
